### PR TITLE
feat(chain): Extend `changes` RPC API to allow querying multiple accounts at once

### LINKED ISSUE #2129: Changes RPC should allow querying multiple accounts at once
Changes RPC API should allow requesting changes for the given list of accounts. Currently it supports requesting change per only one account.

### PR DESCRIPTION
**Pull request by frol**
*Resolves #2129*

I opted to just replace `account_id` input arguments with `account_ids` expecting a list of those. It is technically a breaking change, but since the API is experimental, it should be expected to evolve rapidly.

Documentation request is here: https://github.com/nearprotocol/docs/issues/173

## API

`EXPERIMENTAL_changes` RPC endpoint accepts the following types of requests:

```js
{
    "block_id": /* block hash | block height | omit it if you want to use `finality` */,
    "finality": /* "latest" | "final" | "near-final" | omit it if you want to se `block_id` */,
    "changes_type": /* "account_changes" | "single_access_key_changes" | "all_access_key_changes" | "contract_code_changes" | "data_changes" */,
    /* extra fields based on `changes_type` */
}
```

the extra fields based on `changes_type` are the following:

* ` "account_changes"`:
    * `account_ids`
* `"single_access_key_changes"`:
    * `keys` a list of
        * `account_id`
        * `public_key`
* `"all_access_key_changes"`:
    * `account_ids`
* `"contract_code_changes`:
    * `account_ids`
* `"data_changes"`:
    * `account_ids`
    * `key_prefix_base64` (base64-encoded string of the key-prefix data saved from a smart contract)

### Example of Account changes

When an account gets modified (created, updated, removed), we can query the block which contains the receipts for the transaction:

```js
http post http://localhost:3030 jsonrpc=2.0 id=dontcare \
    method=EXPERIMENTAL_changes \
    'params:={
        "block_id": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW"
        "changes_type": "account_changes",
        "account_ids": ["new-account"]
    }'
```

we observe the following:

```js
{
    "block_hash": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW",
    "changes": [
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "8rPN9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTobgRv"
            },
            "type": "account_update",
            "change": {
                "account_id": "new-account",
                "amount": "0",
                "locked": "0",
                "code_hash": "11111111111111111111111111111111",
                "storage_usage": 100,
                "storage_paid_at": 9
            }
        }
    ]
}
```

If the account was deleted:

```js
{
    "block_hash": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW",
    "changes": [
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "8rPN9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTobgRv"
            },
            "type": "account_deletion",
            "change": {
                "account_id": "new-account"
            }
        }
    ]
}
```

### Example of Data changes

When some key gets modified (created, updated, removed), we can query the block which contains the receipts for the transaction:

```js
http post http://localhost:3030 jsonrpc=2.0 id=dontcare \
    method=EXPERIMENTAL_changes \
    'params:={
        "block_id": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW"
        "changes_type": "data_changes",
        "account_ids": ["new-account"],
        "key_prefix_base64": "bXlfa2V5"  /* base64-encoded "my_key" */
    }'
```

we observe the following (assume the key was modified twice in a single block, and then removed):

```js
{
    "block_hash": "DYHRzRbxUR1ANPdCQcgQE9g5zyYQoDZ1k8BJEQ3hDSgW",
    "changes": [
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "111N9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTob111",
            },
            "type": "data_update",
            "change": {
                "account_id": "new-account",
                "key_base64": "bXlfa2V5",  /* base64-encoded "my_key" */
                "value_base64":  "bXlfdmFsdWVfMQ==" /* base64-encoded "my_value_1" */,
            }
        },
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "222N9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTob222"
            },
            "type": "data_update",
            "change": {
                "account_id": "new-account",
                "key_base64": "bXlfa2V5",  /* base64-encoded "my_key" */
                "value_base64": "bXlfdmFsdWVfMg==" /* base64-encoded "my_value_2" */
            }
        },
        {
            "cause": {
                "type": "receipt_processing",
                "receipt_hash": "333N9tHY9MtT23TqBvvBGBQzwskA5fuZDjmiLpTob333"
            },
            "type": "data_deletion",
            "change": {
                "account_id": "new-account",
                "key_base64": "bXlfa2V5"  /* base64-encoded "my_key" */
            }
        }
    ]
}
```

## Test plan

* [x] Update the existing `pytest` suite (`sanity/rpc_state_changes.py`)
* [x] Add tests for an empty list of account ids
* [x] Add tests for a list of account ids which don't have any changes in the given block
* [x] Add tests for multiple account ids where some accounts don't have any changes in the given block

### CODE REVIEW

File: chain/chain/src/store.rs [hunk 1]
```diff
@@ -887,81 +887,72 @@ impl ChainStoreAccess for ChainStore {
...
-            }
-
-            StateChangesRequest::AllAccessKeyChanges { .. }
-            | StateChangesRequest::DataChanges { .. } => &mut changes_per_key_prefix,
-        };
-
         Ok(match state_changes_request {
-            StateChangesRequest::AccountChanges { account_id, .. } => {
-                StateChanges::from_account_changes(changes_per_key, account_id)?
+            StateChangesRequest::AccountChanges { account_ids } => {
```
Comment (frol) [inline chain/chain/src/store.rs]: FYI, I finally arrived at the state when the code for reading the state changes from the storage and converting those raw values into proper responses is done outside of this function in the right places, so ultimately, it is easier to just "monomorphize" the implementations for each individual type of request (notice how the diff size is +95 -90 meaning that implementing a new feature only required a few extra lines of code).

File: chain/chain/src/store.rs [hunk 2]
```diff
@@ -887,81 +887,72 @@ impl ChainStoreAccess for ChainStore {
...
-            StateChangesRequest::SingleAccessKeyChanges { account_id, .. }
-            | StateChangesRequest::AllAccessKeyChanges { account_id, .. } => {
-                let access_key_pk = match state_changes_request {
-                    StateChangesRequest::SingleAccessKeyChanges { access_key_pk, .. } => {
-                        Some(access_key_pk)
-                    }
-                    _ => None,
-                };
-                StateChanges::from_access_key_changes(changes_per_key, account_id, access_key_pk)?
+            StateChangesRequest::SingleAccessKeyChanges { account_ids, access_key_pk } => {
```
Comment (evgenykuzyakov) [inline chain/chain/src/store.rs]: This is strange API to have from normal perspective. Why would someone would need to track multiple accounts with only one access key. It means those accounts share the same access key. Which is terrible idea from pseudo-anonymity and security perspective.

I'd suggest to change it to provide pairs of `(account_id, access_key_pk)` for the API to make sense.

Comment (frol) [reply chain/chain/src/store.rs]: Very good point! I will update the API.

File: chain/chain/src/store.rs (hunk 1 above)
Comment (frol) [reply chain/chain/src/store.rs]: Well, the PR progressed to +727 -520 :rofl: Well, those are mostly tests, which now cover not only happy paths but also corner cases.

File: pytest/tests/sanity/rpc_state_changes.py [hunk 3]
```diff
@@ -17,407 +17,556 @@
...
+        response = node.get_changes(request)
+        assert 'result' in response, "the request did not succeed: %r" % response
+        response = response['result']
+        diff = deepdiff.DeepDiff(expected_response, response, **kwargs)
+        assert not diff, \
+            "query node #%d same changes gives different results %r (expected VS actual):\n%r\n%r" \
+            % (node_index, diff, expected_response, response)
+
+
+def test_changes_with_new_account():
```
Comment (frol) [inline pytest/tests/sanity/rpc_state_changes.py]: @SkidanovAlex @bowenwang1996 @ailisp BTW, the benefit of having the tests in functions it that you can run the test suite with `pytest` (Python testing framework) and have a nice output for free:

image

**Comment by ailisp** [comment]
I like pytest too
